Updating dev branch with latest code

DurgaMasabattula opened the pull request:
Raising PR to master branch for reviewing

DurgaMasabattula committed:
Updating dev branch with latest code

tatemz commented on src/assets/styles.scss:
```diff
@@ -0,0 +1,7 @@
+@font-face {
+	font-family: 'myfont';
```
Small note - good practice to have consistent naming.

```suggestion
	font-family: 'lato-regular';
```

DurgaMasabattula replied:
Updated this

tatemz commented on src/components/App/App.component.js:
```diff
@@ -0,0 +1,28 @@
+
+class App extends React.Component {
+	constructor(props) {
+		super(props);
```
Unless performing extra steps inside of the `constructor`, it can be left out.

DurgaMasabattula replied:
Removed it as am not performing anything else in constructor()

tatemz commented on src/components/App/App.component.scss:
```diff
@@ -0,0 +1,7 @@
+    position: fixed !important;
+    width: 100% !important;
+    z-index: 9 !important;
+}
```
It's not good practice to abuse the `!important` CSS rule. It's better to add an additional class to the element you are wanting to customize and increase the [selector specificity](https://css-tricks.com/specifics-on-css-specificity/) within the CSS.

DurgaMasabattula replied:
Updated the css with specifity and removed !important

tatemz commented on src/components/ProductDetails/ProductDetails.component.js:
```diff
@@ -0,0 +1,69 @@
+		)
+	}
+}
+
```
Be sure to add the [`propTypes`](https://reactjs.org/docs/typechecking-with-proptypes.html) this component depends on before exporting it.

DurgaMasabattula replied:
added propTypes

tatemz commented on src/components/ProductDetails/ProductDetails.component.js:
```diff
@@ -0,0 +1,69 @@
+import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'react-bootstrap';
+import RatingStarsComponent from '../RatingStars/RatingStars.component';
+
+class ProductDetailsComponent extends React.Component {
```
It is good practice to build stateless components as often as possible in order to increase the testability of a component, and in this case, since we are just showing a modal, we can display this component using a `Route` and pass in the `product` when navigating to this route.

☝️ above is just an opinion, but by adding a `Route` to display product details, one could create shareable URLs that navigate the app directly to that page (e.g. http://localhost:3000/$PRODUCT_ID)

DurgaMasabattula replied:
Converted ProductDetails as a functional component.

API Call to get the product specific details using product ID is not available, so displaying them in Modal instead of new screen.

tatemz commented on src/components/ProductList/ProductList.component.js:
```diff
@@ -0,0 +1,106 @@
+import LoaderComponent from '../Loader/Loader.component';
+
+class ProductListComponent extends React.Component {
+	constructor(props) {
```
Though the `ProductDetailsComponent` could or _could not_ be made stateless, this `ProductListComponent` could most definitely be made stateless. Rather than storing `products`, `totalPages`, `pagesize`, etc, in the state, they could be passed in as props.

Also if `pagesize` never changes, then it should be in `props` and not in `state`.

tatemz replied:
The Redux [`compose`](https://redux.js.org/api/compose) function was created for passing a store/state to component `props`

tatemz replied:
Redux (or any state manager for that matter) was not indicated in the instructions, but if possible, could you refactor this app to use a state manager for the data that is loaded in from the API? That would provide a good example of writing enterprise-sustainable code.

tatemz commented on src/components/RatingStars/RatingStars.component.js:
```diff
@@ -0,0 +1,21 @@
+import './RatingStars.component.scss';
+class RatingStarsComponent extends React.Component {
+	constructor(props) {
+		super(props);
```
Unless performing extra steps inside of the `constructor`, it can be left out.

DurgaMasabattula replied:
Removed constructor, and updated it as functional component

tatemz commented on src/components/ProductDetails/ProductDetails.component.js:
```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+import './ProductDetails.component.scss';
+import endpoints from '../../api/endpoints.config';
+import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'react-bootstrap';
```
Some of these imports are never used. Consider using a `eslint` and `eslint-plugin-react`.

DurgaMasabattula replied:
Updated this, sorry I didnt realize it

tatemz commented on src/components/ProductList/ProductList.component.scss:
```diff
@@ -0,0 +1,3 @@
+.main-content {
+	padding-top: 50px !important;
```
It's not good practice to abuse the `!important` CSS rule. It's better to add an additional class to the element you are wanting to customize and increase the [selector specificity](https://css-tricks.com/specifics-on-css-specificity/) within the CSS.

DurgaMasabattula replied:
removed !important from CSS

tatemz commented on src/components/ProductList/ProductList.component.js:
```diff
@@ -0,0 +1,106 @@
+			<div className="row mx-0 main-content">
+				{ this.state.loader ? <LoaderComponent/> : '' }
+				{rendercomponent} 
+				{ this.state.viewProduct ? <ProductDetailsComponent product={this.state.viewProduct} close={this.emptyProductOnClose}/> : '' }
```
Consider using a new `Route` here

tatemz reviewed (changes requested):
- [ ] Lint code and use consistent indentation throughout
- [ ] Implement a state manager to load and pass API results to the components
- [ ] Use `PropTypes` throughout
- [ ] Add unit and [enzyme](https://airbnb.io/enzyme/) tests (see ["Running Tests"](https://facebook.github.io/create-react-app/docs/running-tests)

tatemz commented:
Thanks for submitting your test project @DurgaMasabattula! Above you will find some comments and suggestions. Don't hesitate to ask any questions if clarity is needed.

DurgaMasabattula committed:
Removing !important from CSS files, using specifity concept to overcome the usage

DurgaMasabattula commented:
@tatemz , thank you for taking time in reviewing it. I will work on the changes :)

DurgaMasabattula committed:
Removed the constructor function at instances of not using it exactly

DurgaMasabattula committed:
RatingStarts to functional component
Loader to functional component
Product details to functional component
PropTypes updated wherever needed

Commit:
Updated font-name to have consistency

DurgaMasabattula commented:
@tatemz , I updated the PR as per the review comments. please take a look at them, thank you for taking time on this.
